File: src/main/java/io/github/chindeaone/collectiontracker/api/tokenapi/TokenFetcher.java
package io.github.chindeaone.collectiontracker.api.tokenapi;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import io.github.chindeaone.collectiontracker.api.URLManager;
import io.github.chindeaone.collectiontracker.util.PlayerData;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;

import static io.github.chindeaone.collectiontracker.api.URLManager.HTTP_CLIENT;

public class TokenFetcher {

    private static final Logger logger = LogManager.getLogger(TokenFetcher.class);

    public String fetchToken() throws Exception {
        URI uri = URI.create(URLManager.TOKEN_URL);

        HttpRequest request = HttpRequest.newBuilder(uri)
                .timeout(Duration.ofSeconds(5))
                .header("X-UUID", PlayerData.INSTANCE.getPlayerUUID())
                .header("User-Agent", URLManager.AGENT)
                .header("Accept", "application/json")
                .GET()
                .build();

        HttpResponse<InputStream> response =
                HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofInputStream());

        int status = response.statusCode();
        if (status != 200) {
            logger.error("[SCT]: Failed to fetch token, response code: {}", status);
            return null;
        }

        try (Reader reader = new InputStreamReader(response.body(), StandardCharsets.UTF_8)) {
            JsonObject jsonResponse = JsonParser.parseReader(reader).getAsJsonObject();
            logger.info("[SCT]: Successfully fetched token");

            return jsonResponse.getAsJsonPrimitive("token").getAsString();
        } catch (Exception e) {
            logger.error("[SCT]: An error occurred while parsing the token response", e);
            return null;
        }
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/api/tokenapi/TokenFetcher.kt
package io.github.chindeaone.collectiontracker.api.tokenapi

import com.google.gson.JsonParser
import io.github.chindeaone.collectiontracker.api.ApiManager
import io.github.chindeaone.collectiontracker.utils.PlayerData
import io.github.chindeaone.collectiontracker.utils.chat.ChatUtils.sendMessage
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.util.concurrent.CompletableFuture

object TokenFetcher {

    private val logger: Logger = LogManager.getLogger(TokenFetcher::class.java)

    fun fetchToken(notify: Boolean): CompletableFuture<String?> {
        return try {
            val serverId = ApiManager.serverId
            if (serverId == null) {
                logger.error("[SCT]: Mojang authentication was not completed.")
                return CompletableFuture.completedFuture(null)
            }

            val headers = mapOf(
                "X-UUID" to PlayerData.playerUUID,
                "X-NAME" to PlayerData.playerName,
                "X-SERVER-ID" to serverId
            )

            ApiManager.requestAsync("token", headers)
                .thenApply { response ->
                    if (response.statusCode() != 200) {
                        logger.error("[SCT]: Failed to fetch token, response code: {}", response.statusCode())

                        if (notify) {
                            sendMessage("§cFailed to fetch token.", true)
                        }

                        null
                    } else {
                        val json = JsonParser.parseString(response.body()).asJsonObject

                        logger.info("[SCT]: Successfully fetched token")
                        if (notify) {
                            sendMessage("§aSuccessfully fetched a new token.", true)
                        }

                        json["token"].asString
                    }
                }.exceptionally { e ->
                    logger.error("[SCT]: Failed to fetch token", e)
                    null
                }
        } catch (e: Exception) {
            logger.error("[SCT]: Failed to fetch token", e)
            return CompletableFuture.completedFuture(null)
        }
    }
}
